src/outcomes.ts

```diff
@@ -0,0 +1,298 @@
+      OutcomeConclusion.indexOf(failRunOn);
+    return didFail ? [{ language, reason }] : [];
+  });
+
+  if (failures.length > 0) {
+    console.log("The following languages did not build successfully:");
+    for (const { language, reason } of failures) {
+      console.log(`${language}: ${reason}`);
+    }
+    return false;
+  }
+
+  return true;
+}
+
+export function categorizeOutcome({
+  outcome,
+  baseOutcome,
+}: {
+  outcome: Outcomes[string];
+  baseOutcome?: Outcomes[string];
+}): {
+  conclusion: OutcomeConclusion;
+  reason: string;
+  isMergeConflict?: boolean;
+  isPending?: boolean;
+} {
+  const commitConclusion = outcome.commit?.completed?.conclusion;
+
+  // If we have old diagnostics, only fail run against new diagnostics.
+  const diagnostics = getNewDiagnostics(
+    outcome.diagnostics,
+    baseOutcome?.diagnostics,
+  );
+  const diagnosticCounts = countDiagnosticLevels(diagnostics);
+
+  // If we have old checks, only fail run against new checks.
+  const checks = getNewChecks(outcome, baseOutcome);
+  const checkFailures = CheckType.filter(
+    (checkType) =>
+      checks[checkType] &&
+      checks[checkType].status === "completed" &&
+      checks[checkType].completed.conclusion !== "success",
+  );
+
+  // Special case: noops and cancels are successful
+  if (commitConclusion === "noop") {
+    return {
+      conclusion: "success",
+      reason: "Code was not generated because the target is skipped.",
+    };
+  }
+  if (commitConclusion === "cancelled") {
+    return {
+      conclusion: "success",
+      reason: "Code was not generated because the build was cancelled.",
+    };
+  }
+
+  // Fatal reasons
+  if (!commitConclusion || commitConclusion === "fatal") {
+    return {
+      conclusion: "fatal",
+      reason: "Code was not generated because there was a fatal error.",
+      isPending: outcome.commit?.status !== "completed",
+    };
+  }
+  if (commitConclusion === "timed_out") {
+    return {
+      conclusion: "fatal",
+      reason: "Timed out.",
+    };
+  }
+  if (
+    ![
+      // Merge conflicts are warnings, not fatal:
+      "merge_conflict",
+      "upstream_merge_conflict",
+      // Success conclusion are handled below:
+      "error",
+      "warning",
+      "note",
+      "success",
+      // All other commit conclusions are unknown, and thus fatal.
+    ].includes(commitConclusion)
+  ) {
+    return {
+      conclusion: "fatal",
+      reason: `Unknown conclusion: ${commitConclusion}`,
+    };
+  }
+  if (diagnosticCounts.fatal > 0) {
+    return {
+      conclusion: "fatal",
+      reason: `Found ${diagnosticCounts.fatal} fatal diagnostics.`,
+    };
+  }
+
+  // Error reasons
+  if (diagnosticCounts.error > 0) {
+    return {
+      conclusion: "error",
+      reason: `Found ${diagnosticCounts.error} error diagnostics.`,
+    };
+  }
+  if (checkFailures.includes("build")) {
+    return {
+      conclusion: "error",
+      reason: "The build CI job failed.",
+    };
+  }
```

Comment:
how hard would it be to provide a link here?

Reply:
not hard at all